src/daft-table/src/ops/pivot.rs

```diff
@@ -0,0 +1,109 @@
+use crate::Table;
+use common_error::{DaftError, DaftResult};
+use daft_core::{
+    array::ops::{as_arrow::AsArrow, IntoGroups},
+    datatypes::UInt64Array,
+    IntoSeries,
+};
+use daft_dsl::ExprRef;
+
+impl Table {
+    pub fn pivot(
+        &self,
+        group_by: ExprRef,
+        pivot_col: ExprRef,
+        values_col: ExprRef,
+        names: Vec<String>,
+    ) -> DaftResult<Table> {
+        let groupby_series = self.eval_expression(&group_by)?;
+        let pivot_series = self.eval_expression(&pivot_col)?;
+        let value_series = self.eval_expression(&values_col)?;
+
+        let (group_key_indices, group_vals_indices) = groupby_series.make_groups()?;
+        let (pivot_key_indices, pivot_vals_indices) = pivot_series.make_groups()?;
+
+        let group_keys_series = {
+            let indices_as_series = UInt64Array::from(("", group_key_indices)).into_series();
+            groupby_series.take(&indices_as_series)?
+        };
+        let pivot_keys_series = {
+            let indices_as_series =
+                UInt64Array::from(("", pivot_key_indices.clone())).into_series();
+            pivot_series.take(&indices_as_series)?
+        };
+        let pivot_names_as_series = pivot_keys_series.to_str_values()?;
+        let pivot_names = pivot_names_as_series.utf8()?.as_arrow();
+        if pivot_names.len() > names.len() {
```

Comment:
Not necessarily, for example in an empty partition, pivot_names will be empty, but we still need to create empty columns for these names otherwise the schema won't match with the other non-empty partitions.